src/lib.rs

```diff
@@ -0,0 +1,763 @@
+            .num_threads(self.thread_pool_size as usize)
+            .build()
+            .unwrap();
+
+        let mut g = Graph {
+            graph_type: self.graph_type.clone(),
+            vertices: vec![],
+            edges: vec![],
+            vertex_key: vec![],
+            edge_key: vec![],
+            neighbour_index: HashMap::new(),
+            vertex_index: HashMap::new(),
+        };
+
+        let directed = match self.graph_type {
+            GraphType::Directed => true,
+            GraphType::Undirected => false,
+        };
+        let has_node_file = self.node_file_path.is_some();
+
+        let sep = self.separate;
+        let edge_file = &self.edge_file_path;
+        // preserve a thread for collecting
+        let worker_count = self.thread_pool_size - 1;
+
+        // 0.Create index for separating file
+        let mut rdr = ReaderBuilder::new()
+            .has_headers(true)
+            .delimiter(sep)
+            .from_path(edge_file)
+            .expect("Please check `edge_file` path");
+        let mut wtr = Cursor::new(vec![]);
+        RandomAccessSimple::create(&mut rdr, &mut wtr);
+        let mut idx = RandomAccessSimple::open(wtr).unwrap();
+        // Obtains key names from file
+        for key in rdr.headers().unwrap().iter() {
+            g.edge_key.push(key.to_string());
+        }
+
+        // 1. Construct channel for threads' communication,
+        //    Transmit Box pointer here to avoid large memory copy.
+        let (tx, rx) = channel();
+
+        // 2. Distribute reading task for threads.
+        let job_size = idx.len() / worker_count + 1;
+        let mut real_worker_count = worker_count;
+        for i in 0..worker_count {
+            let tx = tx.clone();
+            let start_index = job_size * i + 1;
+            // Break distribute when no more jobs
+            if start_index >= idx.len() {
+                real_worker_count = i;
+                break;
+            }
+            let start_pos = idx.get(start_index as u64).unwrap();
+            self.thread_pool.install(move || {
+                read_edge_file_task(
+                    sep,
+                    &edge_file,
+                    start_pos,
+                    job_size,
+                    directed,
+                    has_node_file,
+                    tx,
+                );
+            });
+        }
+
+        // 3. Collecting thread's result.
+        let mut vertices: HashMap<T, Vertex<T>> = HashMap::new();
+        let mut edges: HashMap<T, Vec<Edge<T>>> = HashMap::new();
+        let (tx_collect, rx_collect) = channel();
+        self.thread_pool
+            .install(|| collect_edge_task(has_node_file, real_worker_count, rx, tx_collect));
+        let (vertices_rec, edges_rec) = rx_collect.recv().unwrap();
+        vertices = *vertices_rec;
+        edges = *edges_rec;
+
+        if has_node_file {
+            let node_file_path = self.node_file_path.clone().unwrap();
```

Comment:
These two functions (of reading nodes and edges ) are so similar, please consider reusing some codes.